### PR TITLE
Remove owning raw pointers from Rules base class.

### LINKED ISSUE #14389: Continue removing raw owning pointers from Mantid::Geometry
Followup to pull request #14137. Remove raw owning pointers from Rules.h and RuleItems.cpp

### PR DESCRIPTION
**Pull request by quantumsteve**
This fixes #14389.

This replaces the **owning** raw pointer Rule \* with std::unique_ptr<Rule>. Like in PR #14137, I made the copy constructor and copy assignment operator `protected` and removed empty destructors from the derived classes.

The primary goal is to help developers by clarifying ownership. No release notes.

testing: code review

### CODE REVIEW

File: Framework/Geometry/src/Objects/Rules.cpp [hunk 1]
```diff
@@ -231,19 +236,24 @@ int Rule::makeCNFcopy(Rule *&TopRule)
...
           // hence we have to play games with a second
           // gamma->clone()
-          partReplace = new Union(new Intersection(alpha, gamma),
-                                  new Intersection(beta, gamma->clone()));
+          std::unique_ptr<Rule> tmp1 =
+              Mantid::Kernel::make_unique<Intersection>(std::move(alpha),
+                                                        std::move(gamma));
+          std::unique_ptr<Rule> tmp2 =
+              Mantid::Kernel::make_unique<Intersection>(std::move(beta),
+                                                        std::move(gamma));
```
Comment (SimonHeybrock) [inline Framework/Geometry/src/Objects/Rules.cpp]: I am not sure I understand this. Previously there was `gamma->clone()`, now we call `std::move(gamma)` twice, so the second time it should always contain a `nullptr`. I guess that is not what we want?

File: Framework/Geometry/src/Objects/Rules.cpp [hunk 2]
```diff
@@ -323,34 +333,38 @@ int Rule::makeCNF(Rule *&TopRule)
...
```
Comment (SimonHeybrock) [inline Framework/Geometry/src/Objects/Rules.cpp]: Same `gamma->clone()` question, see above.

File: Framework/Geometry/src/Objects/Rules.cpp (hunk 1 above)
Comment (martyngigg) [inline Framework/Geometry/src/Objects/Rules.cpp]: Yes that doesn't look correct. Nice catch. @quantumsteve I assume your intention here was just to reproduce the previous behaviour?

Comment (quantumsteve) [inline Framework/Geometry/src/Objects/Rules.cpp]: @martyngigg, yes that was my intention. PR #14458 should fix this error.

**Pull request merged**